### PR TITLE
JAMES-2917 Avoid specifying an unbounded count of routing keys

### PR DESCRIPTION
**Pull request by chibenwa**
Specifying too much routing keys can create ridiculously long URL for
ElasticSearch requests that will eventually get rejected.

`414 Request-URI Too Large`

Mitigation: omitting routing keys for searches on more than X mailboxes, X
being a code defined constant. This will result in a search involving all
shards by default.

### CODE REVIEW

File: mailbox/elasticsearch/src/main/java/org/apache/james/mailbox/elasticsearch/search/ElasticSearchSearcher.java [hunk 1]
```diff
@@ -96,18 +97,24 @@ private SearchRequest prepareSearch(Collection<MailboxId> mailboxIds, SearchQuer
...
-            .map(RoutingKey::asString)
-            .toArray(String[]::new);
+    private Optional<String[]> toRoutingKey(Collection<MailboxId> mailboxIds) {
+        if (mailboxIds.size() < MAX_ROUTING_KEY) {
+            return Optional.of(mailboxIds.stream()
+                .map(routingKeyFactory::from)
+                .map(RoutingKey::asString)
+                .toArray(String[]::new));
+        }
+        return Optional.empty();
```
Comment (trantienduchn) [inline mailbox/elasticsearch/src/main/java/org/apache/james/mailbox/elasticsearch/search/ElasticSearchSearcher.java]: does the search result of the same query can be different if I don't specify any routing?

**Comment by mbaechler** [comment]
> Mitigation: omitting routing keys for searches on more than X mailboxes, X
being a code defined constant. This will result in a search involving all
shards by default.

by default or when the number of mailboxes is high?

Is there still cases where the use of routing keys makes an advantage?

File: mailbox/elasticsearch/src/main/java/org/apache/james/mailbox/elasticsearch/search/ElasticSearchSearcher.java [hunk 2]
```diff
@@ -96,18 +97,24 @@ private SearchRequest prepareSearch(Collection<MailboxId> mailboxIds, SearchQuer
...
+            .map(request::routing)
+            .orElse(request);
     }
 
-    private String[] toRoutingKeys(Collection<MailboxId> mailboxIds) {
-        return mailboxIds.stream()
-            .map(routingKeyFactory::from)
-            .map(RoutingKey::asString)
-            .toArray(String[]::new);
+    private Optional<String[]> toRoutingKey(Collection<MailboxId> mailboxIds) {
```
Comment (mbaechler) [inline mailbox/elasticsearch/src/main/java/org/apache/james/mailbox/elasticsearch/search/ElasticSearchSearcher.java]: you probably lack a unit test on SearchRequest preparation

**Comment by chibenwa** [comment]
> by default or when the number of mailboxes is high?

By default yes. Very very very likely when you have a lot of routing key.

> Is there still cases where the use of routing keys makes an advantage?

Clearly yes. When they allow to ignore some of our shards (ie their count iss low) yes. As performance tests demonstrated it.

File: mailbox/elasticsearch/src/main/java/org/apache/james/mailbox/elasticsearch/search/ElasticSearchSearcher.java (hunk 1 above)
Comment (chibenwa) [reply mailbox/elasticsearch/src/main/java/org/apache/james/mailbox/elasticsearch/search/ElasticSearchSearcher.java]: No, as no routing means "all shards" including the one owning your data.

**Comment by Arsnael** [comment]
```
[87623685eb33a114887edf4c6bc5de6465f6fac6] [ERROR] Errors: 
[87623685eb33a114887edf4c6bc5de6465f6fac6] [ERROR]   MemorySpamAssassinContractTest.movingAMailToTrashShouldNotImpactSpamassassinLearning » ConditionTimeout
```
test this please

**Comment by Arsnael** [comment]
```
[259306a8fd890bd686aa2e1c77d8984d523b9611] [ERROR] Failed to execute goal on project event-sourcing-pojo: Could not resolve dependencies for project org.apache.james:event-sourcing-pojo:jar:3.5.0-SNAPSHOT: Failed to collect dependencies at com.google.guava:guava:jar:25.1-jre -> com.google.j2objc:j2objc-annotations:jar:1.1: Failed to read artifact descriptor for com.google.j2objc:j2objc-annotations:jar:1.1: Could not transfer artifact com.google.j2objc:j2objc-annotations:pom:1.1 from/to central (https://repo.maven.apache.org/maven2): /root/.m2/repository/com/google/j2objc/j2objc-annotations/1.1/j2objc-annotations-1.1.pom.part (No such file or directory) -> [Help 1]
```
... test this please

**Comment by chibenwa** [comment]
test this please

**Comment by Arsnael** [comment]
```
[259306a8fd890bd686aa2e1c77d8984d523b9611] [ERROR] Errors: 
[259306a8fd890bd686aa2e1c77d8984d523b9611] [ERROR]   WebAdminUtilsTest.serverShouldBeAbletoStartConcurrently:38 » NotTerminated
```
test this please

File: mailbox/elasticsearch/src/test/java/org/apache/james/mailbox/elasticsearch/search/ElasticSearchSearcherTest.java [hunk 3]
```diff
@@ -128,9 +128,9 @@ void tearDown() throws IOException {
     }
 
     @Test
-    void searchingInFiveMailboxesShouldReturnOnlyTheseMailboxesMessagesUid() throws Exception {
+    void searchingInALargeNumberOfMailboxesShouldReturnAllMailboxesMessagesUid() throws Exception {
         MailboxSession session = MailboxSessionUtil.create(USERNAME);
-        int numberOfMailboxes = 6;
+        int numberOfMailboxes = 700;
```
Comment (chibenwa) [inline mailbox/elasticsearch/src/test/java/org/apache/james/mailbox/elasticsearch/search/ElasticSearchSearcherTest.java]: What's the execution speed of this?

Comment (blackheaven) [reply mailbox/elasticsearch/src/test/java/org/apache/james/mailbox/elasticsearch/search/ElasticSearchSearcherTest.java]: 14 seconds, anyway, it passes without the fix with a lower value

Comment (chibenwa) [reply mailbox/elasticsearch/src/test/java/org/apache/james/mailbox/elasticsearch/search/ElasticSearchSearcherTest.java]: Let s see if it is stable then

**Comment by Arsnael** [comment]
Merged

**Pull request closed without merging**